src/Transports.AspNetCore/Extensions/GraphQLBuilderMiddlewareExtensions.cs

```diff
@@ -1,28 +0,0 @@
-#nullable enable
-
-using GraphQL.DI;
-using GraphQL.Server.Transports.AspNetCore;
-using GraphQL.Types;
-
-namespace GraphQL.Server;
-
-/// <summary>
-/// GraphQL specific extension methods for <see cref="IGraphQLBuilder"/>.
-/// </summary>
-public static class GraphQLBuilderMiddlewareExtensions
```

Comment:
Why did you remove these methods and switched back to approach when middleware is not registered into DI container and does not inherit `IMiddleware`? See #735 and https://github.com/graphql-dotnet/server/pull/735#issuecomment-1043452927

Reply:
Because it requires the use of `AddServer`.  This design, once complete, will allow for configuring and enabling HTTP middleware (both GET/POST and WebSockets) with a single line of code - `UseGraphQL`.  Notice that options are not normally set through DI.  This is by design, so that multiple endpoints with different configurations can be applied, and/or for different schemas.  I think this would be typical in situations where the WebSocket endpoint is different than the GET/POST endpoint (which may be rare), or if connection-level authorization settings are different for different schemas.

It is still possible to configure middleware options through DI; something like this:

```cs
services.AddGraphQL(b => b
    .Services.Configure<GraphQLHttpMiddlewareOptions>(...)
    //other stuff here
    );

app.UseGraphQL<GraphQLHttpMiddleware<ISchema>>();
```

I also made sure the design works with custom options classes for custom middleware.

```cs
app.UseGraphQL<MyMiddlware>(new MyMiddlewareOptions { /* ... */ });
```

There is one unfortunate restriction when not using DI -- and that is that only the first constructor of `GraphQLHttpMiddleware<TSchema>` will ever be used.  I did not run into a major problem with that restriction, but the end result of my WebSocket design is that if a user wants to have a derived WebSocket handling class, it takes some extra setup and derived classes.  This is partially due to the layers of abstraction I have set up, which you will see in the subsequent PR adding the new WebSocket support.

Reply:
Note that according to MS docs, the convention-based pattern is effectively a singleton, so that's good.

Reply:
> Notice that options are not normally set through DI. 

Regarding this and other notes I agree. I thought about the same.

> is effectively a singleton

🤔 Indeed. I looked through the MS `UseMiddleware` code one more time and see. So my changes with `AddMiddleware` methods were poinless.

> only the first constructor of `GraphQLHttpMiddleware<TSchema>` will ever be used.

Why so? It depends on `args` passed in `UseMiddleware`:
```csharp
public static IApplicationBuilder UseMiddleware(this IApplicationBuilder app, Type middleware, params object[] args)
        {
            if (typeof(IMiddleware).GetTypeInfo().IsAssignableFrom(middleware.GetTypeInfo()))
            {
                // IMiddleware doesn't support passing args directly since it's
                // activated from the container
                if (args.Length > 0)
                {
                    throw new NotSupportedException(Resources.FormatException_UseMiddlewareExplicitArgumentsNotSupported(typeof(IMiddleware)));
                }

                return UseMiddlewareInterface(app, middleware);
            }

            var applicationServices = app.ApplicationServices;
            return app.Use(next =>
            {
                var methods = middleware.GetMethods(BindingFlags.Instance | BindingFlags.Public);
                var invokeMethods = methods.Where(m =>
                    string.Equals(m.Name, InvokeMethodName, StringComparison.Ordinal)
                    || string.Equals(m.Name, InvokeAsyncMethodName, StringComparison.Ordinal)
                    ).ToArray();

                if (invokeMethods.Length > 1)
                {
                    throw new InvalidOperationException(Resources.FormatException_UseMiddleMutlipleInvokes(InvokeMethodName, InvokeAsyncMethodName));
                }

                if (invokeMethods.Length == 0)
                {
                    throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNoInvokeMethod(InvokeMethodName, InvokeAsyncMethodName, middleware));
                }

                var methodInfo = invokeMethods[0];
                if (!typeof(Task).IsAssignableFrom(methodInfo.ReturnType))
                {
                    throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNonTaskReturnType(InvokeMethodName, InvokeAsyncMethodName, nameof(Task)));
                }

                var parameters = methodInfo.GetParameters();
                if (parameters.Length == 0 || parameters[0].ParameterType != typeof(HttpContext))
                {
                    throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNoParameters(InvokeMethodName, InvokeAsyncMethodName, nameof(HttpContext)));
                }

                var ctorArgs = new object[args.Length + 1];
                ctorArgs[0] = next;
                Array.Copy(args, 0, ctorArgs, 1, args.Length);
                var instance = ActivatorUtilities.CreateInstance(app.ApplicationServices, middleware, ctorArgs);
                if (parameters.Length == 1)
                {
                    return (RequestDelegate)methodInfo.CreateDelegate(typeof(RequestDelegate), instance);
                }

                var factory = Compile<object>(methodInfo, parameters);

                return context =>
                {
                    var serviceProvider = context.RequestServices ?? applicationServices;
                    if (serviceProvider == null)
                    {
                        throw new InvalidOperationException(Resources.FormatException_UseMiddlewareIServiceProviderNotAvailable(nameof(IServiceProvider)));
                    }

                    return factory(instance, context, serviceProvider);
                };
            });
        }
```

Reply:
I said that based on actual results. Nevertheless, as I guessed, it’s using ActivatorUtilities.CreateInstance. I know there’s an attribute you can use to set which constructor to use and stuff. But it won’t intelligently select a constructor based on what services are registered.

The fact that it’s effectively a singleton is documented, not just based on the source. I can find the link if you like.

Reply:
> But it won’t intelligently select a constructor based on what services are registered.

I see. Not a big problem I think. We just follow common design of ASP.NET framework.

> I can find the link if you like.

Yes please.

Reply:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-6.0

> This topic describes how to write convention-based middleware.

> Middleware should follow the [Explicit Dependencies Principle](https://docs.microsoft.com/en-us/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#explicit-dependencies) by exposing its dependencies in its constructor. Middleware is constructed once per application lifetime.

Reply:
> So my changes with AddMiddleware methods were poinless.

lol - well, actually I think I first made the change in the middleware sample in the GraphQL.NET repo, and then you did the same in this repo.  So you could say that both of our changes were pointless - lol.

Reply:
True.

Reply:
After reading the docs I see that convention-based middleware seems to be more convenient than factory-based.

Reply:
Well, in this case it is.  But if someone wanted scoped middleware then factory based would be better.